### PR TITLE
Don't allow invalid template combinations

### LINKED ISSUE #56314: Do not allow breaking v2 index templates by removing in-use component templates
A major design goal of v2 index templates is to allow us to perform more thorough validation at template creation time. Such templates are composed of explicit component templates, so we can validate the resulting template. However, removing a component template that is used by a v2 index template would break any v2 index templates using that component template, thus ruining our plan for more thorough validation of v2 index templates. We should prevent the breakage by not allowing removing component templates that are in use by any v2 index template.

### PR DESCRIPTION
**Pull request by dakrone**
This commit removes the ability to put V2 index templates that reference missing component templates.
It also prevents removing component templates that are being referenced by an existing V2 index
template.

Relates to #53101
Resolves #56314

### CODE REVIEW

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
Looks like CI ran out of memory:
```
12:21:05 java.lang.OutOfMemoryError: Java heap space
12:21:05 Dumping heap to java_pid3691.hprof ...
12:21:08 Heap dump file created [169278035 bytes in 2.711 secs]
```

@elasticmachine run elasticsearch-ci/1

**Comment by dakrone** [comment]
@elasticmachine update branch

**Comment by dakrone** [comment]
@elasticmachine update branch

**Review by andreidan — APPROVED** [review]
LGTM, thanks for fixing this Lee

Left a super minor nit, feel free to ignore

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 1]
```diff
@@ -291,6 +293,33 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
             });
     }
 
+    /**
+     * Validates that the given component template is not used by any index
+     * templates, throwing an error if it is still in use
+     */
+    static void validateNotInUse(Metadata metadata, String templateWildcard) {
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:300]: what do you think about renaming `templateWildcard` to `nameOrWilldcard` (or `templateNameOrWildcard`)

**Pull request merged**